training/src/main/scala/com/airbnb/aerosolve/training/AdditiveModelTrainer.scala

```diff
@@ -73,13 +81,25 @@ object AdditiveModelTrainer {
 
     val paramsBC = sc.broadcast(params)
     var model = modelInitialization(sc, transformed, params)
-    for (i <- 1 to iterations) {
+    var loss = Double.MaxValue
+    for (i <- 1 to iterations
+         if loss >= params.loss.minLoss) {
       log.info(s"Iteration $i")
+      val sgdParams = SgdParams(
+        paramsBC,
+        sc.accumulator(0),
+        sc.accumulator(0))
       val modelBC = sc.broadcast(model)
-      model = sgdTrain(transformed, paramsBC, modelBC)
+      model = sgdTrain(transformed, sgdParams, modelBC)
       modelBC.unpersist()
-
-      TrainingUtils.saveModel(model, output)
+      val newLoss = sgdParams.loss.value/sgdParams.exampleCount.value
+      if (params.loss.useBestLoss && newLoss < loss) {
```

Comment:
Typically using the relevance metric (like the loss here) is not a good idea (relevance loss is an approximate for business metric and best relevance loss does not usually give you the best business metric value). Directly using business metric is a better alternative.